# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmInstance.h
# Commit: d6506b5ee07b
# Full Hash: d6506b5ee07b5c65f889ce7c2214b0172f205898
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2021-09-16 09:43:03
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Introduce indirect stubs to optimize call_indirect. r=lth
#   
#   This patch introduces indirect stubs.
#   An indirect stub is a stub that takes care of any switching activities needed for call_indirect.
#   Before this patch, we have to conservatively assume that any call_indirect's target can be from a foreign instance,
# ==============================================================================

diff -r 80a164c1054e -r d6506b5ee07b js/src/wasm/WasmInstance.h
--- a/js/src/wasm/WasmInstance.h	Wed Sep 15 11:02:57 2021 +0000
+++ b/js/src/wasm/WasmInstance.h	Wed Sep 15 11:10:15 2021 +0000
@@ -66,6 +66,8 @@
   const void** addressOfTypeId(const TypeIdDesc& typeId) const;
   FuncImportTls& funcImportTls(const FuncImport& fi);
   TableTls& tableTls(const TableDesc& td) const;
+  void* checkedCallEntry(const Tier tier, const uint32_t functionIndex) const;
+  void* checkedCallEntryForWasmImportedFunction(JSFunction* fun) const;
 
   // Only WasmInstanceObject can call the private trace function.
   friend class js::WasmInstanceObject;
@@ -98,6 +100,8 @@
                        uint8_t* nextPC,
                        uintptr_t highestByteVisitedInPrevFrame);
 
+  Instance* getOriginalInstanceAndFunction(uint32_t funcIdx, JSFunction** fun);
+
   JS::Realm* realm() const { return realm_; }
   const Code& code() const { return *code_; }
   const CodeTier& code(Tier t) const { return code_->codeTier(t); }
@@ -173,6 +177,16 @@
                                uint32_t dstOffset, uint32_t srcOffset,
                                uint32_t len);
 
+  void* getIndirectStub(uint32_t funcIndex, TlsData* targetTlsData) const;
+  void* createIndirectStub(uint32_t funcIndex, TlsData* targetTlsData);
+  bool createManyIndirectStubs(const VectorOfIndirectStubTarget& targets,
+                               const Tier tier);
+  bool ensureIndirectStubs(const Uint32Vector& elemFuncIndices,
+                           uint32_t srcOffset, uint32_t len, const Tier tier,
+                           const bool tableIsImportedOrExported);
+  bool ensureIndirectStub(FuncRef* ref, const Tier tier,
+                          const bool tableIsImportedOrExported);
+
   // Debugger support:
 
   JSString* createDisplayURL(JSContext* cx);